Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignature.swift

```diff
@@ -16,130 +16,313 @@ import SymbolKit
+                    // Ignore whitespace in text tokens. Here we use a loop instead of `filter` to avoid a potential temporary allocation.
+                    for char in spelling where char != space {
+                        accumulated.append(char)
+                    }
+                    
+                default:
+                    continue
+                }
+            } else {
+                switch fragment.kind {
+                case .identifier where preciseIdentifier != nil,
+                     .typeIdentifier,
+                     .text:
+                    let spelling = fragment.spelling.utf8
+                    
+                    // Ignore whitespace. Here we use a loop instead of `filter` to avoid a potential temporary allocation.
+                    for char in spelling where char != space {
+                        accumulated.append(char)
+                    }
+                    
+                default:
+                    continue
+                }
             }
         }
         
-        accumulated.removeAll(where: \.isWhitespace)
-        
-        if accumulated.first == ":" {
-            _ = accumulated.removeFirst()
+        // Check if the type names are wrapped in redundant parenthesis and remove them
+        if accumulated.first == openParen, accumulated.last == closeParen, !accumulated[...].isTuple() {
+            // In case there are multiple
+            // Use a temporary slice until all the layers of redundant parenthesis have been removed.
+            var temp = accumulated[...]
+            
+            repeat {
+                temp = temp.dropFirst().dropLast()
+            } while temp.first == openParen && temp.last == closeParen && !temp.isTuple()
+            
+            // Adjust the markers so that they align with the expected characters
+            let difference = (accumulated.count - temp.count) / 2
+            
+            accumulated = .init(temp)
+            
+            for index in markers.indices {
+                markers[index] -= difference
+                
+                assert(accumulated[markers[index]] == uppercaseA || accumulated[markers[index]] == uppercaseD  || accumulated[markers[index]] == uppercaseO, """
+                Unexpectedly found '\(String(cString: [accumulated[index], 0]))' at \(index) which should be either an Array, Optional, or Dictionary marker in \(String(cString: accumulated + [0]))
+                """)
+            }
         }
         
-        while accumulated.first == "(", accumulated.last == ")", !accumulated.isTuple() {
-            // Remove extra layers of parenthesis unless the type is a tuple
-            accumulated = accumulated.dropFirst().dropLast()
+        assert(markers.allSatisfy { [uppercaseA, uppercaseD, uppercaseO].contains(accumulated[$0]) }, """
+        Unexpectedly found misaligned markers: \(markers.map { "(index: \($0), char: \(String(cString: [accumulated[$0], 0])))" })
+        """)
```

Comment:
Assertions only run in debug builds so it's okay if their computations have a performance impact. 

Because of this, the pattern of `assert(list.allSatisfy(predicate)` is better than a for loop of assertions because the empty loop (which may get optimized away) doesn't remain when the assertion is removed.

Similarly, the message of the assertion is only evaluated if the assertion condition fails, which means that it can be as expensive as it wants to be—in order to surface more relevant information to the developer—because the program will trap immediately afterwards anyway.

Reply:
> Assertions only run in debug builds so it's okay if their computations have a performance impact.

I wasn't aware of this, TIL! 🎉